File: sudoq-app/sudoqmodel/src/test/java/de/sudoq/model/sudoku/PositionMapTests.java
package de.sudoq.model.sudoku;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.util.BitSet;

import org.junit.Test;

import de.sudoq.model.sudoku.Position;
import de.sudoq.model.sudoku.PositionMap;

public class PositionMapTests {

	@Test
	public void testStandardUsage() {
		PositionMap<BitSet> map = new PositionMap<>(Position.get(9, 9));
		BitSet b = new BitSet();
		map.put(Position.get(3, 2), b);
		b.set(7);
		assertTrue(map.get(Position.get(3, 2)).get(7));
		b.clear();
		assertEquals(map.get(Position.get(3, 2)).cardinality(), 0);

		PositionMap<BitSet> map2 = map.clone();
		assertEquals(map.get(Position.get(3, 2)), map2.get(Position.get(3, 2)));
		assertEquals(map.get(Position.get(2, 3)), map2.get(Position.get(2, 3)));
	}

	@Test
	public void testIllegalArguments() {

		try {
			new PositionMap<BitSet>(Position.get(1, 0));
			fail("No IllegalArgumentException was thrown, altough the dimension was null");
		} catch (IllegalArgumentException e) {
		}

		try {
			new PositionMap<BitSet>(Position.get(0, 1));
			fail("No IllegalArgumentException was thrown, altough the dimension was null");
		} catch (IllegalArgumentException e) {
		}

		PositionMap<BitSet> map = new PositionMap<>(Position.get(9, 9));

		try {
			map.put(Position.get(10, 9), new BitSet());
			fail("No IllegalArgumentException was thrown, altough the Position was invalid");
		} catch (IllegalArgumentException e) {
		}

		try {
			map.get(Position.get(10, 9));
			fail("No IllegalArgumentException was thrown, altough the Position was invalid");
		} catch (IllegalArgumentException e) {
		}

		try {
			map.get(Position.get(9, 10));
			fail("No IllegalArgumentException was thrown, altough the Position was invalid");
		} catch (IllegalArgumentException e) {
		}
	}

}

File: sudoq-app/sudoqmodel/src/test/kotlin/de/sudoq/model/sudoku/PositionMapTests.kt
package de.sudoq.model.sudoku


import org.amshove.kluent.invoking
import org.amshove.kluent.`should be`
import org.amshove.kluent.`should be true`
import org.amshove.kluent.`should throw`
import org.junit.jupiter.api.Test
import java.util.BitSet

class PositionMapTests {

    @Test
    fun standardUsage() {
        val map = PositionMap.Builder<BitSet>(Position[9, 9])
        val b = BitSet()
        map.put(Position[3, 2], b)
        b.set(7)
        (map.build()[Position[3, 2]][7]).`should be true`()
        b.clear()
        map.build()[Position[3, 2]].cardinality().`should be`(0)
    }

    @Test
    fun illegalArguments() {
        invoking {
            PositionMap.Builder<BitSet>(Position[1, 0])
        }.`should throw`(IllegalArgumentException::class)

        invoking {
            PositionMap.Builder<BitSet>(Position[0, 1])
        }.`should throw`(IllegalArgumentException::class)

        val map = PositionMap.Builder<BitSet>(Position[9, 9])
        invoking {
            map.put(Position[10, 9], BitSet())
        }.`should throw` (IllegalArgumentException::class)

        invoking {
            map.build()[Position[10, 9]]
        }.`should throw`(IllegalArgumentException::class)

        invoking {
            map.build()[Position[9, 10]]
        }.`should throw`(IllegalArgumentException::class)

    }
}